use rez api

nobsclement opened the pull request:
Hello,

This is the PR I spoke to you about, that changes some code to use rez api.
I was able to compile multiple variants of USD successfully using this new version.

The tool behaviour is a bit different with this new code.

Since it uses rez api, the tool uses rez config's `packages_path`. You're not limited to the `install_prefix` directory anymore for dependency resolution.
If not specified, the `install_prefix` is now the local_packages_path from rez config.
You don't have to specifiy every constraints anymore, rez-cook will just get the highest possible version available that satisfies all the other constraints.
You can use rez request syntax for the recipe and constraints args.
The cooked variant uses pinned versions. For example, cooking USD:
Before:
`usd-22.08/platform-windows/arch-AMD64/vs-2022/cfg-release/boost-1.79.0/tbb/openexr-2.4|3.1.2+/ocio-2.2.0/oiio/python-3.9.12/ptex/osd`
After:
`usd-22.08\platform-windows\arch-AMD64\vs-2022\cfg-release\boost-1.79.0\tbb-2020.3\openexr-3.1.5\ocio-2.2.0\oiio-2.3.16.0\python-3.9.12\ptex-2.4.1\osd-3.4.4`

I tested as best as I could but there might be some bugs. I couldn't test on linux, would like to know if it works correctly over there :)
Feel free to give me any feedback.

Some commands to test things out:
Build usd with vfxrp-2022: `rez python .\rez-cook.py --prefix D:\rez-cooked-packages -d usd -c vfxrp-2022`
Try to build usd with vfxrp-2022 and python-3.7: `rez python .\rez-cook.py --prefix D:\rez-cooked-packages -d usd -c vfxrp-2022 python-3.7`
That should output an error:
```
[ERROR] Failed to resolve context: ['vfxrp-2022', 'python-3.7', 'usd']
[ERROR] The following package conflicts occurred: (python-3.9 <--!--> python==3.7.9)
[ERROR] Resolve paths starting from initial requests to conflict:
  vfxrp-2022 --> [vfxrp==2022 --> python-3.9
  python-3.7 --> [python==3.7.9
  usd --> usd --> python-3.9
```

nobsclement committed:
use rez api

nobsclement committed:
don't build missing packages from constraints

That may happen when you want to constrain a recipe to vfxrp for example.

nobsclement committed:
validate contexts earlier and fix vfxrp like constraints

nobsclement committed:
better constraints handling if a variant require is not a package require

For example, if you want to build a usd package with alembic compatibility.
You may have a usd package that has an optional alembic variant that allows you to build usd with or without alembic support if its recipe supports it as an optional variant.

anderslanglands commented:
Just got around to testing this. I moved my old packages out the way, did a `rez-bind os` and then `rez-python ./rez-cook.py oiio-2.3 -c cfg-release cxx11abi-0 --prefix=$HOME/packages-rc` and got:
```bash
anders@LAPTOP-0S80KQL8:~/code/rez-cook-cduranti$ rez-python ./rez-cook.py oiio-2.3 -c cfg-release cxx11abi-0 --prefix=$HOME/packages-rc

Using already installed dependencies:
    cfg-release
    platform-linux
    arch-x86_64
    cxx11abi-0/platform-linux/arch-x86_64
    cmake-3.23.2/platform-linux/arch-x86_64
    python-3.10.5/platform-linux/arch-x86_64
    boost-1.79.0/platform-linux/arch-x86_64/cxx11abi/python/cfg
    gif-5.2.1/platform-linux/arch-x86_64/cxx11abi/cfg
    imath-3.1.5/platform-linux/arch-x86_64/cxx11abi/cfg
    jpegturbo-2.0.6/platform-linux/arch-x86_64/cxx11abi/cfg
    zlib-1.2.12/platform-linux/arch-x86_64/cxx11abi/cfg
    openexr-3.1.5/platform-linux/arch-x86_64/cxx11abi/cfg
    pybind11-2.8.1/platform-linux/arch-x86_64/cxx11abi/python/cfg
    ocio-2.2.0/platform-linux/arch-x86_64/cxx11abi/cfg/openexr/python (dde7c8f0ef1c00338327693070f66a695f0e7a56)
    openjpeg-2.4.0/platform-linux/arch-x86_64/cxx11abi/cfg
    png-1.6.35/platform-linux/arch-x86_64/cxx11abi/cfg
    ptex-2.4.1/platform-linux/arch-x86_64/cxx11abi/cfg
    raw-0.21.0/platform-linux/arch-x86_64/cxx11abi/cfg
    tbb-2020.3/platform-linux/arch-x86_64
    tiff-4.3.0/platform-linux/arch-x86_64/cxx11abi/cfg
    webp-1.1.0/platform-linux/arch-x86_64/cxx11abi/cfg
    oiio-2.3.16.0/platform-linux/arch-x86_64/cxx11abi/cfg/boost/tbb/openexr/ocio/python/ptex (ff742266c76d7c8d344fe5332518fe39501eb08e)
Nothing to cook
```

nobsclement commented:
I suspect that you have all those packages installed somewhere else in one of your `$REZ_PACKAGES_PATH`. Since the fork uses the rez API, it now looks for cooked recipes using the same code as when you do a `rez-env` or `rez-search` which look at every `$REZ_PACKAGES_PATH`.

What's the output of `rez-search oiio-2.3`? If it finds a package, that's why there is nothing to cook!

How about: `REZ_PACKAGES_PATH=$HOME/packages-rc && rez-python ./rez-cook.py oiio-2.3 -c cfg-release cxx11abi-0 --prefix=$HOME/packages-rc`
With this snippet, as long as `$HOME/packages-rc` is missing some packages from the recipe, it should cook something.

anderslanglands commented:
Yeah that's the confusing thing: if I do `rez-search oiio` I get `No
matching family found`. Could it be getting confused with the recipes
somehow?

On Fri, 9 Sept 2022 at 20:23, Clément Duranti ***@***.***>
wrote:

> I suspect that you have all those packages installed somewhere else in one
> of your $REZ_PACKAGES_PATH. Since the fork uses the rez API, it now looks
> for cooked recipes using the same code as when you do a rez-env or
> rez-search which look at every $REZ_PACKAGES_PATH.
>
> What's the output of rez-search oiio-2.3? If it finds a package, that's
> why there is nothing to cook!
>
> How about: REZ_PACKAGES_PATH=$HOME/packages-rc && rez-python
> ./rez-cook.py oiio-2.3 -c cfg-release cxx11abi-0 --prefix=$HOME/packages-rc
> With this snippet, as long as $HOME/packages-rc is missing some packages
> from the recipe, it should cook something.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/anderslanglands/rez-cook/pull/1#issuecomment-1241662171>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOYQXORSQ22ZMBBETVXCJ3V5LXWNANCNFSM56SZPE3A>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

anderslanglands commented:
That's kinda what it looks like given that the dependencies it's found
don't have versions on any of the variants

On Fri, 9 Sept 2022 at 20:50, Anders Langlands ***@***.***>
wrote:

> Yeah that's the confusing thing: if I do `rez-search oiio` I get `No
> matching family found`. Could it be getting confused with the recipes
> somehow?
>
> On Fri, 9 Sept 2022 at 20:23, Clément Duranti ***@***.***>
> wrote:
>
>> I suspect that you have all those packages installed somewhere else in
>> one of your $REZ_PACKAGES_PATH. Since the fork uses the rez API, it now
>> looks for cooked recipes using the same code as when you do a rez-env or
>> rez-search which look at every $REZ_PACKAGES_PATH.
>>
>> What's the output of rez-search oiio-2.3? If it finds a package, that's
>> why there is nothing to cook!
>>
>> How about: REZ_PACKAGES_PATH=$HOME/packages-rc && rez-python
>> ./rez-cook.py oiio-2.3 -c cfg-release cxx11abi-0 --prefix=$HOME/packages-rc
>> With this snippet, as long as $HOME/packages-rc is missing some packages
>> from the recipe, it should cook something.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/anderslanglands/rez-cook/pull/1#issuecomment-1241662171>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAOYQXORSQ22ZMBBETVXCJ3V5LXWNANCNFSM56SZPE3A>
>> .
>> You are receiving this because you commented.Message ID:
>> ***@***.***>
>>
>

anderslanglands commented:
I just tried setting REZ_RECIPES_PATH (it was not set in my previous
attempt) and that doesn't change anything.

On Fri, 9 Sept 2022 at 20:51, Anders Langlands ***@***.***>
wrote:

> That's kinda what it looks like given that the dependencies it's found
> don't have versions on any of the variants
>
> On Fri, 9 Sept 2022 at 20:50, Anders Langlands ***@***.***>
> wrote:
>
>> Yeah that's the confusing thing: if I do `rez-search oiio` I get `No
>> matching family found`. Could it be getting confused with the recipes
>> somehow?
>>
>> On Fri, 9 Sept 2022 at 20:23, Clément Duranti ***@***.***>
>> wrote:
>>
>>> I suspect that you have all those packages installed somewhere else in
>>> one of your $REZ_PACKAGES_PATH. Since the fork uses the rez API, it now
>>> looks for cooked recipes using the same code as when you do a rez-env
>>> or rez-search which look at every $REZ_PACKAGES_PATH.
>>>
>>> What's the output of rez-search oiio-2.3? If it finds a package, that's
>>> why there is nothing to cook!
>>>
>>> How about: REZ_PACKAGES_PATH=$HOME/packages-rc && rez-python
>>> ./rez-cook.py oiio-2.3 -c cfg-release cxx11abi-0 --prefix=$HOME/packages-rc
>>> With this snippet, as long as $HOME/packages-rc is missing some
>>> packages from the recipe, it should cook something.
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/anderslanglands/rez-cook/pull/1#issuecomment-1241662171>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AAOYQXORSQ22ZMBBETVXCJ3V5LXWNANCNFSM56SZPE3A>
>>> .
>>> You are receiving this because you commented.Message ID:
>>> ***@***.***>
>>>
>>

nobsclement commented:
Hmm, are the recipes themselves on your `$REZ_PACKAGES_PATH`? It shouldn't be a problem for most recipes except the ones that have explicit variants baked in the recipe (like `vs` for example).

Could it be a linux specific behaviour that I've not caught because I've only been able to test on windows?

Can you add those two lines right above the print at line 536 and try again with `--debug`?
``` python
if args.debug:
    variant_text = f"{variant_text} ({dep.base})"
```
So that it looks like:
``` python
    if selected_installed:
        print()
        print("Using already installed dependencies:")
        for dep in selected_installed.values():
            dirs = [x.safe_str() for x in dep.variant_requires]
            subpath_str = os.path.join(*dirs) if dirs else None
            variant_text = ""
            if dep.subpath:
                variant_text = f"{os.path.sep}{subpath_str}"
                if dep.hashed_variants:
                    variant_text = f"{variant_text} ({dep.subpath})"
            if args.debug:
                variant_text = f"{variant_text} ({dep.base})"
            print(f"    {dep.qualified_package_name}{variant_text}")
```

It should help debug the problem.

I can push a small change to make sure the `$REZ_RECIPES_PATH` is not in the packages path list, but I'm not sure if it's a good idea or not. Having the the recipes path in `$REZ_PACKAGES_PATH` should not happen and if it's the case, it means your rezconfig is wrongly configured. That's how I see it at least, you should have the final word on that!

nobsclement commented:
Just tried on WSL2 Ubuntu, fresh rez, fresh clone of my rez-cook fork and fresh clone of your recipes repo.
``` bash
cdu@cdu:~/projects/rez-cook$ export REZ_RECIPES_PATH=~/projects/rez-recipes
cdu@cdu:~/projects/rez-cook$ rez-python ./rez-cook.py oiio-2.3 -c cfg-release cxx11abi-0 --prefix=$HOME/packages-rc

Using already installed dependencies:
    platform-linux
    arch-x86_64

Cooking:
    cfg-release
    cxx11abi-0/platform-linux/arch-x86_64
    cmake-3.23.2/platform-linux/arch-x86_64
    python-3.10.5/platform-linux/arch-x86_64
    boost-1.79.0/platform-linux/arch-x86_64/cxx11abi-0/python-3.10.5/cfg-release
    gif-5.2.1/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    imath-3.1.5/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    jpegturbo-2.0.6/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    zlib-1.2.12/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    openexr-3.1.5/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    pybind11-2.8.1/platform-linux/arch-x86_64/cxx11abi-0/python-3.10.5/cfg-release
    ocio-2.2.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release/openexr-3.1.5/python-3.10.5 (945fa61ae1b7b6853447aa7c1a8fe7ede393a5d9)
    openjpeg-2.4.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    png-1.6.35/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    ptex-2.4.1/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    raw-0.21.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    tbb-2020.3/platform-linux/arch-x86_64
    tiff-4.3.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    webp-1.1.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
    oiio-2.3.16.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release/boost-1.79.0/tbb-2020.3/openexr-3.1.5/ocio-2.2.0/python-3.10.5/ptex-2.4.1 (675de28de34c9661bd546c844d60eff3ae2b929c)

Proceed? (y/n): n
Exiting
```

Here is my rez packages path:

```bash
cdu@cdu:~/projects/rez-cook$ rez config packages_path
- /home/cdu/packages
- /home/cdu/.rez/packages/int
- /home/cdu/.rez/packages/ext
```

I just realised that you don't have specific version in your variants on the log you sent earlier:
Yours: `ocio-2.2.0/platform-linux/arch-x86_64/cxx11abi/cfg/openexr/python`
Mine: `ocio-2.2.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release/openexr-3.1.5/python-3.10.5`

Looks like yours has the old behaviour, which should not happen on my fork. I'd say that something is wrong with either your rezconfig or that maybe you're not using my code?

anderslanglands commented:
Definitely using your code, I cloned a fresh copy of the repo.

With that debug print I get:

***@***.***:~/code/rez-cook-cduranti$ rez python ./rez-cook.py
oiio --debug

Using already installed dependencies:
    platform-linux (/home/anders/packages/platform/linux)
    arch-x86_64 (/home/anders/packages/arch/x86_64)
    cfg-release (/mnt/c/Users/anders/code/rez-recipes/cfg/release)
    cmake-3.23.2/platform-linux/arch-x86_64
(/mnt/c/Users/anders/code/rez-recipes/cmake/3.23.2)
    cxx11abi-1/platform-linux/arch-x86_64
(/mnt/c/Users/anders/code/rez-recipes/cxx11abi/1)
    python-3.10.5/platform-linux/arch-x86_64
(/mnt/c/Users/anders/code/rez-recipes/python/3.10.5)
    boost-1.79.0/platform-linux/arch-x86_64/cxx11abi/python/cfg
(/mnt/c/Users/anders/code/rez-recipes/boost/1.79.0)
    gif-5.2.1/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/gif/5.2.1)
    imath-3.1.5/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/imath/3.1.5)
    jpegturbo-2.0.6/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/jpegturbo/2.0.6)
    zlib-1.2.12/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/zlib/1.2.12)
    openexr-3.1.5/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/openexr/3.1.5)
    pybind11-2.8.1/platform-linux/arch-x86_64/cxx11abi/python/cfg
(/mnt/c/Users/anders/code/rez-recipes/pybind11/2.8.1)
    ocio-2.2.0/platform-linux/arch-x86_64/cxx11abi/cfg/openexr/python
(dde7c8f0ef1c00338327693070f66a695f0e7a56)
(/mnt/c/Users/anders/code/rez-recipes/ocio/2.2.0)
    openjpeg-2.4.0/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/openjpeg/2.4.0)
    png-1.6.35/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/png/1.6.35)
    ptex-2.4.1/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/ptex/2.4.1)
    raw-0.21.0/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/raw/0.21.0)
    tbb-2020.3/platform-linux/arch-x86_64
(/mnt/c/Users/anders/code/rez-recipes/tbb/2020.3)
    tiff-4.3.0/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/tiff/4.3.0)
    webp-1.1.0/platform-linux/arch-x86_64/cxx11abi/cfg
(/mnt/c/Users/anders/code/rez-recipes/webp/1.1.0)

oiio-2.3.16.0/platform-linux/arch-x86_64/cxx11abi/cfg/boost/tbb/openexr/ocio/python/ptex
(ff742266c76d7c8d344fe5332518fe39501eb08e)
(/mnt/c/Users/anders/code/rez-recipes/oiio/2.3.16.0)
Nothing to cook

Does that mean it is interpreting my recipes path as a package path? Here's
my config:

***@***.***:~/code/rez-cook-cduranti$ rez config packages_path
- /home/anders/packages
- /home/anders/.rez/packages/int
- /home/anders/.rez/packages/ext

***@***.***:~/code/rez-cook-cduranti$ ls ~/packages
arch  os  platform

On Fri, 9 Sept 2022 at 22:28, Clément Duranti ***@***.***>
wrote:

> Just tried on WSL2 Ubuntu, fresh rez, fresh clone of my rez-cook fork and
> fresh clone of your recipes repo.
>
> ***@***.***:~/projects/rez-cook$ export REZ_RECIPES_PATH=~/projects/rez-recipes
> ***@***.***:~/projects/rez-cook$ rez-python ./rez-cook.py oiio-2.3 -c cfg-release cxx11abi-0 --prefix=$HOME/packages-rc
>
> Using already installed dependencies:
>     platform-linux
>     arch-x86_64
>
> Cooking:
>     cfg-release
>     cxx11abi-0/platform-linux/arch-x86_64
>     cmake-3.23.2/platform-linux/arch-x86_64
>     python-3.10.5/platform-linux/arch-x86_64
>     boost-1.79.0/platform-linux/arch-x86_64/cxx11abi-0/python-3.10.5/cfg-release
>     gif-5.2.1/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     imath-3.1.5/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     jpegturbo-2.0.6/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     zlib-1.2.12/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     openexr-3.1.5/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     pybind11-2.8.1/platform-linux/arch-x86_64/cxx11abi-0/python-3.10.5/cfg-release
>     ocio-2.2.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release/openexr-3.1.5/python-3.10.5 (945fa61ae1b7b6853447aa7c1a8fe7ede393a5d9)
>     openjpeg-2.4.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     png-1.6.35/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     ptex-2.4.1/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     raw-0.21.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     tbb-2020.3/platform-linux/arch-x86_64
>     tiff-4.3.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     webp-1.1.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release
>     oiio-2.3.16.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release/boost-1.79.0/tbb-2020.3/openexr-3.1.5/ocio-2.2.0/python-3.10.5/ptex-2.4.1 (675de28de34c9661bd546c844d60eff3ae2b929c)
>
> Proceed? (y/n): n
> Exiting
>
> Here is my rez packages path:
>
> ***@***.***:~/projects/rez-cook$ rez config packages_path
> - /home/cdu/packages
> - /home/cdu/.rez/packages/int
> - /home/cdu/.rez/packages/ext
>
> I just realised that you don't have specific version in your variants on
> the log you sent earlier:
> Yours: ocio-2.2.0/platform-linux/arch-x86_64/cxx11abi/cfg/openexr/python
> Mine:
> ocio-2.2.0/platform-linux/arch-x86_64/cxx11abi-0/cfg-release/openexr-3.1.5/python-3.10.5
>
> Looks like yours has the old behaviour, which should not happen on my
> fork. I'd say that something is wrong with either your rezconfig or that
> maybe you're not using my code?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/anderslanglands/rez-cook/pull/1#issuecomment-1241796075>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOYQXLHREZPIHECUPUWNITV5MGNPANCNFSM56SZPE3A>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Commit:
remove recipes path if it's in rez packages path

If we don't remove it, recipes with baked variants (like the vs recipe) will be flagged as being cooked already.

nobsclement commented:
I pushed a small fix to force remove the recipes path from packages path just in case.
I have absolutely no idea what's going on on your side though :/ I can't reproduce the behaviour you're having.

Assuming it's a bug in the code, it could be either:
* The recipes are considered as being packages without recipes (https://github.com/cduranti/rez-cook/blob/0464babc93919e273b348976c934bdb54cb2dff8/rez-cook.py#L436-L458)
* The recipes are variant of some installed packages (https://github.com/cduranti/rez-cook/blob/0464babc93919e273b348976c934bdb54cb2dff8/rez-cook.py#L462-L520)

Looking at your config and the fact that none of the recipes are cooked, those lines should trigger for uncooked recipes: https://github.com/cduranti/rez-cook/blob/0464babc93919e273b348976c934bdb54cb2dff8/rez-cook.py#L476-L479
At that point, `installed_packages` should be `[]` for the recipes that are not cooked and `[Package(...)]` for cooked recipes/found packages.

I'd love to debug it and understand what's going on, do you have any suggestion on how I could reproduce?

Commit:
actually check if recipes path is in packages path before trying to remove it

anderslanglands commented:
OK so this is weird. In my WSL setup I have the rez-rezipes repo on my Windows disk, and then I symlinked it into WSL. If I delete the symlink and check out a fresh copy of the recipes repo, everything works. I have no idea why this would be the case. Do you?

Sookhaal committed:
get the real path of recipes path if it's a symlink

Sookhaal commented:
I was able to reproduce and the fix was quite simple. I did not think about symlinks at all :/

nobsclement commented:
I was able to reproduce and the fix was quite simple. I did not think about symlinks at all :/

anderslanglands commented:
Hah well that was simple in the end :) I’ll take another look over the code tomorrow to make sure I understand it then merge. Thanks!

nobsclement committed:
fix error when recipes path is not a symlink
